config/set_config.py

```diff
@@ -0,0 +1,148 @@
+    else:
+        raise KeyError(f"Missing required key: '{field_name}'")
+
+
+# main action
+def set_config(env: str):
+    file_dir = os.path.dirname(os.path.abspath(__file__))
+    env_dir = os.path.join(file_dir, env)
+
+    print(f"Setting config for environment: {env}")
+    print(f"Looking for configs in: {env_dir}")
+
+    # load settings.json
+    settings_path = os.path.join(env_dir, "settings.json")
+    if not os.path.isfile(settings_path):
+        raise FileNotFoundError(f"Missing settings.json at {settings_path}")
+    settings = load_json_file(settings_path)
+    print(f"Loaded settings:\n{settings}")
+
+    # extract settings.json attributes
+    url = extract_attribute(settings, "url")
+    api_key = extract_attribute(settings, "api_key")
+    headers = {"X-API-Key": api_key}
+
+    # load and apply config.json
+    config_path = os.path.join(env_dir, "config.json")
+    if not os.path.isfile(config_path):
+        raise FileNotFoundError(f"Missing config.json at {config_path}")
+    config = load_json_file(config_path)
+    print(f"Applying default config:\n{config}")
+
+    try:
+        resp = requests.put(url, json=config, headers=headers)
+        resp.raise_for_status()
+    except Exception as e:
+        print(f"Skipping {os.path.basename(config_path)} due to error: {e}")
+
+    # find and apply any ip-specific configs
+    ip_config_paths = glob.glob(os.path.join(env_dir, "ip_*.config.json"))
+    for ip_config_path in ip_config_paths:
+        basename = os.path.basename(ip_config_path)
+
+        try:
+            ip, ip_config = parse_ip_config_file(ip_config_path)
+            print(f"Applying IP-specific config from {basename}:\n{ip_config}")
+
+            resp = requests.put(f"{url}?IP={ip}", json=ip_config, headers=headers)
+            resp.raise_for_status()
+        except requests.exceptions.RequestException as e:
+            print(f"Skipping {basename}, failed to send HTTP request: {e}")
+        except ValueError as e:
+            print(f"Skipping {basename}, invalid IP address format: {e}")
+        except Exception as e:
+            print(f"Skipping {basename} due to error: {e}")
+
+
+# args parser
+def main():
+    parser = argparse.ArgumentParser(description="Set environment configuration.")
+    parser.add_argument(
+        "--retry",
+        action="store_true",
+        help="If setting config fails, wait 1 minute and retry indefinitely until successful.",
+    )
+    parser.add_argument(
+        "env",
+        type=str,
+        help="The environment to configure (e.g., dev, qa, preprod, prod).",
+    )
+
+    args = parser.parse_args()
+
+    while True:
+        try:
+            set_config(args.env)
+            print(f"Configuration for '{args.env}' set successfully.")
+            break
+        except requests.exceptions.RequestException as e:
+            print(f"HTTP request failed: {e}")
+        except FileNotFoundError as e:
+            print(f"Missing file: {e}")
```

Comment:
Fatal error if the file is settings.json or config.json but not if its `ip_*.config.json`